src/pybind/matrix/vector_pybind.cc

```diff
@@ -44,20 +45,66 @@ void pybind_vector(py::module& m) {
-            v->Data(),        // ptr
-            py::none());      // pass a base object to avoid copy!
-        // (fangjun): the base object can be anything containing a non-null
-        // ptr. I cannot think of a better way than to pass a `None` object.
-        //
-        // `numpy` instead of `Numpy`, `ToNumpy` or `SomeOtherName` is used here
-        // because we want to follow the style in PyKaldi and PyTorch using the
-        // method `numpy()` to convert a matrix/tensor to a numpy array.
+      .def("numpy",
+           [](VectorBase<float>* v) {
+             return py::array_t<float>(
+                 {v->Dim()},       // shape
+                 {sizeof(float)},  // stride in bytes
+                 v->Data(),        // ptr
+                 py::none());      // pass a base object to avoid copy!
+             // (fangjun): the base object can be anything containing a non-null
+             // ptr. I cannot think of a better way than to pass a `None`
+             // object.
+             //
+             // `numpy` instead of `Numpy`, `ToNumpy` or `SomeOtherName` is used
+             // here because we want to follow the style in PyKaldi and PyTorch
+             // using  the  method `numpy()` to convert a matrix/tensor to a
+             // numpy array.
+           })
+      .def("to_dlpack", [](VectorBase<float>* v) {
+        // we use the name `to_dlpack` because PyTorch uses the same name
+
+        // the created `managed_tensor` will be freed in
+        // `DLManagedTensorDeleter`, so no memory leak here.
+        auto* managed_tensor = new DLManagedTensor();
+        managed_tensor->manager_ctx = nullptr;
+
+        // setup the deleter to free allocated memory.
+        // refer to
+        // https://github.com/pytorch/pytorch/blob/master/torch/csrc/Module.cpp#L361
+        // for how and when the deleter is invoked.
+        managed_tensor->deleter = &DLManagedTensorDeleter;
+
+        auto* tensor = &managed_tensor->dl_tensor;
+        tensor->data = v->Data();
+        tensor->ctx.device_type = kDLCPU;
+        tensor->ctx.device_id = 0;
+
+        tensor->ndim = 1;
+
+        tensor->dtype.code = kDLFloat;
+        tensor->dtype.bits = 32;  // single precision float
+        tensor->dtype.lanes = 1;
+
+        // `shape` and `strides` are freed in `DLManagedTensorDeleter`, so
+        // no memory leak here .
+        tensor->shape = new int64_t[1];
+        tensor->shape[0] = v->Dim();
+
+        tensor->strides = new int64_t[1];
+        tensor->strides[0] = 1;
+        tensor->byte_offset = 0;
+
+        // WARNING(fangjun): the name of the capsule MUST be `dltensor` for
+        // PyTorch; refer to
+        // https://github.com/pytorch/pytorch/blob/master/torch/csrc/Module.cpp#L383/
+        // for more details
+        return py::capsule(managed_tensor, "dltensor");
       });
 
+  // TODO(fangjun): add `from_dlpack` once kaldi::Vector is changed to borrow a
+  // ptr from the outside
```

Comment:
@csukuangfj the issue with adding from_dlpack to Vector is it won't work for CuVector due to different memory allocators.  Would be better, I think, to add it to SubVector, and ideally somehow have the memory be managed by the Python object?  [when we wrap these things, I assume there must be some kind of Python object that contains the C++ object?]

Reply:
I agree.

Reply:
now SubVector/SubMatrix/CuSubVector/CuSubMatrix support both
`from_dlpack` and `to_dlpack`.

Vector/Matrix/CuVector/CuMatrix support ONLY `to_dlpack`